File: src/audio/worklet/envelope.ts
// Attack/release envelope follower, run once per analysis hop (not per
// sample). Fast attack + slow release makes band energies bloom and decay
// like the sound instead of flickering per-frame.
export class EnvelopeFollower {
  private value = 0
  private initialized = false
  private readonly attackCoeff: number
  private readonly releaseCoeff: number

  constructor(attackMs: number, releaseMs: number, hopMs: number) {
    this.attackCoeff = Math.exp(-hopMs / attackMs)
    this.releaseCoeff = Math.exp(-hopMs / releaseMs)
  }

  update(input: number): number {
    // Seed from the first real input rather than climbing from 0 — without
    // this, any paired fast/slow envelope pair (build detector, break
    // detector) reads a spurious "rise" for the first several seconds of
    // playback as the slower one lags behind catching up from zero.
    if (!this.initialized) {
      this.value = input
      this.initialized = true
      return this.value
    }
    const coeff = input > this.value ? this.attackCoeff : this.releaseCoeff
    this.value = coeff * this.value + (1 - coeff) * input
    return this.value
  }
}

// Raw band/flux magnitudes have no fixed scale (depends on input gain,
// source loudness, file mastering). This tracks a slowly-decaying running
// peak and normalizes against it, so values settle into a usable 0..1 range
// without a manual calibration step, adapting over tens of seconds if the
// input's overall level changes.
export class AdaptiveNormalizer {
  private peak = 1e-6
  private readonly decayCoeff: number

  constructor(decayMs: number, hopMs: number) {
    this.decayCoeff = Math.exp(-hopMs / decayMs)
  }

  normalize(value: number): number {
    this.peak = Math.max(value, this.peak * this.decayCoeff)
    return this.peak > 1e-6 ? Math.min(1, value / this.peak) : 0
  }
}

File: crates/hyst-audio/src/dsp/envelope.rs
//! Ported from `src/audio/worklet/envelope.ts`.

/// Attack/release envelope follower, run once per analysis hop.
pub struct EnvelopeFollower {
    value: f32,
    initialized: bool,
    attack_coeff: f32,
    release_coeff: f32,
}

impl EnvelopeFollower {
    pub fn new(attack_ms: f32, release_ms: f32, hop_ms: f32) -> Self {
        Self {
            value: 0.0,
            initialized: false,
            attack_coeff: (-hop_ms / attack_ms).exp(),
            release_coeff: (-hop_ms / release_ms).exp(),
        }
    }

    pub fn update(&mut self, input: f32) -> f32 {
        // Seed from the first real input rather than climbing from 0 — see
        // the TS original's comment: avoids a spurious "rise" reading for
        // the first several seconds of playback on any paired fast/slow
        // envelope (build/break detectors).
        if !self.initialized {
            self.value = input;
            self.initialized = true;
            return self.value;
        }
        let coeff = if input > self.value {
            self.attack_coeff
        } else {
            self.release_coeff
        };
        self.value = coeff * self.value + (1.0 - coeff) * input;
        self.value
    }
}

/// Tracks a slowly-decaying running peak and normalizes against it, so raw
/// magnitudes with no fixed scale settle into 0..1 without manual
/// calibration.
pub struct AdaptiveNormalizer {
    peak: f32,
    decay_coeff: f32,
}

impl AdaptiveNormalizer {
    pub fn new(decay_ms: f32, hop_ms: f32) -> Self {
        Self {
            peak: 1e-6,
            decay_coeff: (-hop_ms / decay_ms).exp(),
        }
    }

    pub fn normalize(&mut self, value: f32) -> f32 {
        self.peak = value.max(self.peak * self.decay_coeff);
        if self.peak > 1e-6 {
            (value / self.peak).min(1.0)
        } else {
            0.0
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn envelope_seeds_from_first_input_not_zero() {
        let mut env = EnvelopeFollower::new(10.0, 300.0, 10.0);
        assert_eq!(env.update(0.5), 0.5);
    }

    #[test]
    fn envelope_attacks_faster_than_it_releases() {
        let mut fast_release = EnvelopeFollower::new(10.0, 10.0, 10.0);
        let mut slow_release = EnvelopeFollower::new(10.0, 500.0, 10.0);
        fast_release.update(1.0);
        slow_release.update(1.0);
        let f = fast_release.update(0.0);
        let s = slow_release.update(0.0);
        assert!(s > f, "slower release should decay less per step");
    }

    #[test]
    fn normalizer_settles_toward_1_at_the_running_peak() {
        let mut norm = AdaptiveNormalizer::new(6000.0, 10.0);
        for _ in 0..50 {
            norm.normalize(1.0);
        }
        assert!((norm.normalize(1.0) - 1.0).abs() < 1e-3);
    }

    #[test]
    fn normalizer_never_exceeds_1() {
        let mut norm = AdaptiveNormalizer::new(6000.0, 10.0);
        norm.normalize(0.1);
        assert!(norm.normalize(10.0) <= 1.0);
    }
}
